src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,154 @@
+                return false;
+            }
+
+            var candidateReplacementNode = SyntaxFactory.IdentifierName("var")
+                                                .WithLeadingTrivia(typeName.GetLeadingTrivia())
+                                                .WithTrailingTrivia(typeName.GetTrailingTrivia());
+
+            var candidateIssueSpan = typeName.Span;
+
+            // If there exists a type named var, return.
+            var conflict = semanticModel.GetSpeculativeSymbolInfo(typeName.SpanStart, candidateReplacementNode, SpeculativeBindingOption.BindAsTypeOrNamespace).Symbol;
+            if (conflict?.IsKind(SymbolKind.NamedType) == true)
+            {
+                return false;
+            }
+
+            if (typeName.Parent.IsKind(SyntaxKind.VariableDeclaration) &&
+                typeName.Parent.IsParentKind(SyntaxKind.LocalDeclarationStatement, SyntaxKind.ForStatement, SyntaxKind.UsingStatement))
+            {
+                var variableDeclaration = (VariableDeclarationSyntax)typeName.Parent;
+
+                // implicitly typed variables cannot be constants.
+                if ((variableDeclaration.Parent as LocalDeclarationStatementSyntax)?.IsConst == true)
+                {
+                    return false;
+                }
+
+                var variable = variableDeclaration.Variables.Single();
+                if (AssignmentSupportsStylePreference(variable.Identifier, typeName, variable.Initializer, semanticModel, optionSet, cancellationToken))
+                {
+                    issueSpan = candidateIssueSpan;
+                }
+            }
+            else if (typeName.IsParentKind(SyntaxKind.ForEachStatement))
+            {
+                issueSpan = candidateIssueSpan;
+            }
+
+            return issueSpan != default(TextSpan);
+        }
+
+        /// <summary>
+        /// Analyzes the assignment expression and rejects a given declaration if it is unsuitable for implicit typing.
+        /// </summary>
+        /// <returns>
+        /// false, if implicit typing cannot be used.
+        /// true, otherwise.
+        /// </returns>
+        protected override bool AssignmentSupportsStylePreference(SyntaxToken identifier, TypeSyntax typeName, EqualsValueClauseSyntax initializer, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken)
+        {
+            var expression = initializer.Value;
+
+            // var cannot be assigned null
+            if (expression.IsKind(SyntaxKind.NullLiteralExpression))
+            {
+                return false;
+            }
+
+            // cannot use implicit typing on method group, anonymous function or on dynamic
+            var declaredType = semanticModel.GetTypeInfo(typeName, cancellationToken).Type;
+            if (declaredType != null
+                && (declaredType.TypeKind == TypeKind.Delegate
+                || declaredType.TypeKind == TypeKind.Dynamic))
+            {
```

Comment:
this would be a lot clearer (to me) as:

```
if (declaredType != null &&
    (... || ...))
{
}

Right now hte alignment makes it seem as if the && and || are part o the same expressoin.